internal/datasystem/fdv2_datasystem.go

```diff
@@ -0,0 +1,302 @@
+		dataStoreStatus:  internal.NewBroadcaster[interfaces.DataStoreStatus](),
+		flagChangeEvent:  internal.NewBroadcaster[interfaces.FlagChangeEvent](),
+	}
+
+	fdv2 := &FDv2{
+		store:                    store,
+		loggers:                  clientContext.GetLogging().Loggers,
+		broadcasters:             bcasters,
+		dataSourceStatusProvider: &dataStatusProvider{},
+	}
+
+	// Yay circular reference.
+	fdv2.dataSourceStatusProvider.system = fdv2
+
+	dataStoreUpdateSink := datastore.NewDataStoreUpdateSinkImpl(bcasters.dataStoreStatus)
+	clientContextCopy := *clientContext
+	clientContextCopy.DataStoreUpdateSink = dataStoreUpdateSink
+	clientContextCopy.DataDestination = store
+	clientContextCopy.DataSourceStatusReporter = fdv2
+
+	cfg, err := cfgBuilder.Build(clientContextCopy)
+	if err != nil {
+		return nil, err
+	}
+
+	fdv2.initializers = cfg.Initializers
+	fdv2.primarySync = cfg.Synchronizers.Primary
+	fdv2.secondarySync = cfg.Synchronizers.Secondary
+	fdv2.disabled = disabled
+
+	if cfg.Store != nil && !disabled {
+		// If there's a persistent Store, we should provide a status monitor and inform Store that it's present.
+		fdv2.dataStoreStatusProvider = datastore.NewDataStoreStatusProviderImpl(cfg.Store, dataStoreUpdateSink)
+		store.WithPersistence(cfg.Store, cfg.StoreMode, fdv2.dataStoreStatusProvider)
+	} else {
+		// If there's no persistent Store, we still need to satisfy the SDK's public interface of having
+		// a data Store status provider. So we create one that just says "I don't know what's going on".
+		fdv2.dataStoreStatusProvider = datastore.NewDataStoreStatusProviderImpl(noStatusMonitoring{}, dataStoreUpdateSink)
+	}
+
+	return fdv2, nil
+}
+
+type noStatusMonitoring struct{}
+
+func (n noStatusMonitoring) IsStatusMonitoringEnabled() bool {
+	return false
+}
+
+func (f *FDv2) Start(closeWhenReady chan struct{}) {
+	if f.disabled {
+		f.loggers.Infof("Data system is disabled, SDK will return application-defined default values")
+		close(closeWhenReady)
+		return
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	f.cancel = cancel
+	f.launchTask(func() {
+		f.run(ctx, closeWhenReady)
+	})
+}
+
+func (f *FDv2) launchTask(task func()) {
+	f.wg.Add(1)
+	go func() {
+		defer f.wg.Done()
+		task()
+	}()
+}
+
+func (f *FDv2) run(ctx context.Context, closeWhenReady chan struct{}) {
+	payloadVersion := f.runInitializers(ctx, closeWhenReady)
+
+	if f.dataStoreStatusProvider.IsStatusMonitoringEnabled() {
+		f.launchTask(func() {
+			f.runPersistentStoreOutageRecovery(ctx, f.dataStoreStatusProvider.AddStatusListener())
+		})
+	}
+
+	f.runSynchronizers(ctx, closeWhenReady, payloadVersion)
+}
+
+func (f *FDv2) runPersistentStoreOutageRecovery(ctx context.Context, statuses <-chan interfaces.DataStoreStatus) {
+	for {
+		select {
+		case newStoreStatus := <-statuses:
+			if newStoreStatus.Available {
+				// The Store has just transitioned from unavailable to available
+				if newStoreStatus.NeedsRefresh {
```

Comment:
Cleaner to make compound conditional instead? Accepting this would require another change below, and probably some formatting fixes.

```suggestion
			if newStoreStatus.Available && newStoreStatus.NeedsRefresh {
				// The Store has just transitioned from unavailable to available
```